hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/LatencyTracker.java

```diff
@@ -0,0 +1,215 @@
+
+
+/**
+ * {@code LatencyTracker} keeps track of service latencies observed by {@code AbfsClient}. Every request adds
+ * its information (success/failure, latency etc) to the {@code LatencyTracker}'s queue.
+ * When a request is made, we check {@code LatencyTracker} to see if there are any latency numbers to be reported.
+ * If there are any, the stats are added to an HTTP header ({@code x-ms-abfs-client-latency}) on the next request. *
+ */
+public class LatencyTracker {
+
+  // the logger
+  private static final Logger LOG = LoggerFactory.getLogger(LatencyTracker.class);
+
+  // the queue to hold latency information
+  private final ConcurrentLinkedQueue<String> queue = new ConcurrentLinkedQueue<String>();
+
+  // whether the latency tracker has been enabled
+  private boolean enabled = false;
+
+  // the host name
+  private String hostName;
+
+  // the file system name
+  private String filesystemName;
+
+  // the account name
+  private String accountName;
+
+  // singleton latency reporting format
+  private String singletonLatencyReportingFormat;
+
+  // aggregate latency reporting format
+  private String aggregateLatencyReportingFormat;
+
+  public LatencyTracker(String filesystemName, String accountName, AbfsConfiguration configuration) {
+    this(filesystemName, accountName, configuration.shouldTrackLatency());
+  }
+
+  protected LatencyTracker(String filesystemName, String accountName, boolean enabled) {
+    this.enabled = enabled;
+    this.filesystemName = filesystemName;
+    this.accountName = accountName;
+
+    LOG.debug("LatencyTracker configuration: {}", enabled);
+
+    if (enabled) {
+      try {
+        hostName = InetAddress.getLocalHost().getHostName();
+      } catch (UnknownHostException e) {
+        hostName = "UnknownHost";
+      }
+
+      singletonLatencyReportingFormat = "h=" + hostName + " t=%s a=" + accountName + " c=" + filesystemName + " cr=%s ce=%s r=%s l=%s%s";
+      aggregateLatencyReportingFormat = "h=" + hostName + " t=%s a=" + accountName + " c=" + filesystemName + " cr=%s ce=%s r=%s l=%s ls=%s lc=%s%s";
+    }
+  }
+
+  public void recordClientLatency(
+          Instant operationStart,
+          String callerName,
+          String calleeName,
+          boolean success,
+          AbfsHttpOperation res) {
+    if (!enabled) {
+      return;
+    }
+
+    Instant operationStop = getLatencyInstant();
+
+    recordClientLatency(operationStart, operationStop, callerName, calleeName, success, res);
+  }
+
+  public void recordClientLatency(
+          Instant operationStart,
+          Instant operationStop,
+          String callerName,
+          String calleeName,
+          boolean success,
+          AbfsHttpOperation res) {
+    if (!enabled) {
+      return;
+    }
+
+    Instant trackerStart = Instant.now();
+    long latency = isValidInstant(operationStart) && isValidInstant(operationStop)
+            ? Duration.between(operationStart, operationStop).toMillis() : -1;
```

Comment:
Is it possible for us to have multiple operations happening at the same time?
For example:
op0 happens from t=3 to t=5
op1 happens from t=4 to t=6
The total time might be 4 but it should be 3.

Reply:
By design, we want that all operations, whether overlapping or not, be separately and independently tracked.